pkg/storage/driver/sql.go

```diff
@@ -0,0 +1,463 @@
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package driver // import "helm.sh/helm/v3/pkg/storage/driver"
+
+import (
+	"fmt"
+	"sort"
+	"strings"
+	"time"
+
+	"github.com/jmoiron/sqlx"
+	migrate "github.com/rubenv/sql-migrate"
+
+	// Import pq for postgres dialect
+	_ "github.com/lib/pq"
+
+	storageerrors "k8s.io/helm/pkg/storage/errors"
+
+	rspb "helm.sh/helm/v3/pkg/release"
+)
+
+var _ Driver = (*SQL)(nil)
+
+var labelMap = map[string]struct{}{
+	"modifiedAt": {},
+	"createdAt":  {},
+	"version":    {},
+	"status":     {},
+	"owner":      {},
+	"name":       {},
+}
+
+var supportedSQLDialects = map[string]struct{}{
+	"postgres": {},
+}
+
+// SQLDriverName is the string name of this driver.
+const SQLDriverName = "SQL"
+
+const sqlReleaseTableName = "releases_v1"
+
+const (
+	sqlReleaseTableKeyColumn        = "key"
+	sqlReleaseTableTypeColumn       = "type"
+	sqlReleaseTableBodyColumn       = "body"
+	sqlReleaseTableNameColumn       = "name"
+	sqlReleaseTableNamespaceColumn  = "namespace"
+	sqlReleaseTableVersionColumn    = "version"
+	sqlReleaseTableStatusColumn     = "status"
+	sqlReleaseTableOwnerColumn      = "owner"
+	sqlReleaseTableCreatedAtColumn  = "createdAt"
+	sqlReleaseTableModifiedAtColumn = "modifiedAt"
+)
+
+const (
+	sqlReleaseDefaultOwner = "helm"
+	sqlReleaseDefaultType  = "helm.sh/release.v1"
+)
+
+// SQL is the sql storage driver implementation.
+type SQL struct {
+	db        *sqlx.DB
+	namespace string
+
+	Log func(string, ...interface{})
+}
+
+// Name returns the name of the driver.
+func (s *SQL) Name() string {
+	return SQLDriverName
+}
+
+func (s *SQL) ensureDBSetup() error {
+	// Populate the database with the relations we need if they don't exist yet
+	migrations := &migrate.MemoryMigrationSource{
+		Migrations: []*migrate.Migration{
+			{
+				Id: "init",
+				Up: []string{
+					`
+						CREATE TABLE releases_v1 (
+							key VARCHAR(67),
+							type VARCHAR(64) NOT NULL,
+							body TEXT NOT NULL,
+							name VARCHAR(64) NOT NULL,
+							namespace VARCHAR(64) NOT NULL,
+						  	version INTEGER NOT NULL,
+							status TEXT NOT NULL,
+							owner TEXT NOT NULL,
+							createdAt INTEGER NOT NULL,
+							modifiedAt INTEGER NOT NULL DEFAULT 0,
+							PRIMARY KEY(key, namespace)
+						);
+						CREATE INDEX ON releases_v1 (name, namespace);
+						CREATE INDEX ON releases_v1 (version);
+						CREATE INDEX ON releases_v1 (status);
+						CREATE INDEX ON releases_v1 (owner);
+						CREATE INDEX ON releases_v1 (createdAt);
+						CREATE INDEX ON releases_v1 (modifiedAt);
+						
+						GRANT ALL ON releases_v1 TO PUBLIC;
+
+						ALTER TABLE releases_v1 ENABLE ROW LEVEL SECURITY;
```

Comment:
Would it be better to create this string first (e.g. `fmt.Sprintf`) before assigning to `Up`  and you could use the `sqlReleaseTableName` definition?

Reply:
You're absolutely right. I've updated the migration to make use of all the constants declared at the top of the file.

Reply:
Indices need updating